test/e2e/nodeportlocal_test.go

```diff
@@ -0,0 +1,408 @@
+		if err != nil || testPodIP.ipv4 == nil {
+			return false, nil
+		}
+
+		ann := pod.GetAnnotations()
+		t.Logf("Got annotations %v for Pod with IP %v", ann, testPodIP.ipv4.String())
+		nplAnn, found = ann[k8s.NPLAnnotationKey]
+		return found, nil
+	})
+	r.NoError(err, "Poll for Pod check failed")
+	return nplAnn, testPodIP.ipv4.String()
+}
+
+func checkNPLRulesForPod(t *testing.T, data *TestData, r *require.Assertions, nplAnnotations []k8s.NPLAnnotation, antreaPod, podIP string, present bool) {
+	var rules []nplRuleData
+	for _, ann := range nplAnnotations {
+		rule := nplRuleData{
+			nodePort: ann.NodePort,
+			podIP:    podIP,
+			podPort:  ann.PodPort,
+		}
+		rules = append(rules, rule)
+	}
+	checkForNPLRuleInIPTables(t, data, r, antreaPod, rules, present)
+}
+
+func checkForNPLRuleInIPTables(t *testing.T, data *TestData, r *require.Assertions, antreaPod string, rules []nplRuleData, present bool) {
+	cmd := []string{"iptables", "-t", "nat", "-S"}
+	t.Logf("Verifying iptables rules %v", rules)
+	err := wait.Poll(time.Second, defaultTimeout, func() (bool, error) {
+		stdout, _, err := data.runCommandFromPod(antreaNamespace, antreaPod, agentContainerName, cmd)
+		if err != nil {
+			t.Logf("Error while checking rules in iptables: %v", err)
+			// Retry, as sometimes error can occur due to concurrent operations on iptables.
+			return false, nil
+		}
+		for _, rule := range rules {
+			ruleSpec := []string{
+				"-p", "tcp", "-m", "tcp", "--dport",
+				fmt.Sprint(rule.nodePort), "-j", "DNAT", "--to-destination", fmt.Sprintf("%s:%d", rule.podIP, rule.podPort),
+			}
+			rs := strings.Join(append([]string{"-A", "ANTREA-NODE-PORT-LOCAL"}, ruleSpec...), " ")
+			t.Logf("Searching for iptables rule: %v", rs)
+
+			if strings.Contains(stdout, rs) && present {
+				t.Logf("Found rule in iptables")
+			} else if !strings.Contains(stdout, rs) && !present {
+				t.Logf("Rule not found in iptables")
+			} else {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	r.NoError(err, "Poll for iptables rules check failed")
+}
+
+func checkTrafficForNPL(t *testing.T, data *TestData, r *require.Assertions, nplAnnotations []k8s.NPLAnnotation, clientName string) {
+	for i := range nplAnnotations {
+		err := data.runNetcatCommandFromTestPod(clientName, nplAnnotations[i].NodeIP, int32(nplAnnotations[i].NodePort))
+		r.NoError(err, "Traffic test failed for NodeIP: %s, NodePort: %d", nplAnnotations[i].NodeIP, nplAnnotations[i].NodePort)
+	}
+}
+
+func enableNPLInConfigmap(t *testing.T, data *TestData) {
+	if err := data.mutateAntreaConfigMap(func(data map[string]string) {
+		antreaAgentConf, _ := data["antrea-agent.conf"]
+		antreaAgentConf = strings.Replace(antreaAgentConf, "#  NodePortLocal: false", "  NodePortLocal: true", 1)
+		antreaAgentConf = strings.Replace(antreaAgentConf, "#nplPortRange: 40000-41000", "nplPortRange: 40000-41000", 1)
+		t.Logf("Updated antreaAgentConf after enabling NodePortLocal: %v", antreaAgentConf)
```

Comment:
same comment as for `updateNPLPortRangeInConfigmap`, please do not log the whole configMap

Reply:
Removed